### PR TITLE
fix geotagging location selection, providers, lifecycle and staleness

### PR DESCRIPTION
**Pull request by thomasbuilds**
Follow-up to #637 and the remaining part of #530. The one-line fix in #637 was incomplete on its own: with the comparison direction corrected, the one-sided staleness check still lets a significantly older but accurate fix win whenever it appears after a fresher one in a candidate list (last known location scans, batched updates), and the wall-clock time comparison it relies on isn't reliable across providers in the first place.

1. **Pick the most accurate recent location for geotagging instead of the least accurate one**: the accuracy tiebreaker was inverted (smaller `getAccuracy()` is better), the staleness check was one-sided (a significantly *older* candidate fell through to the accuracy comparison, making the result order-dependent), and ages were compared via wall-clock `Location.getTime()`, which the platform documentation says "should not be used to order or compare locations". The selection is now: take the newest fix (by `elapsedRealtimeNanos`), and among fixes not significantly older than it, pick the most accurate; missing accuracy ranks last, ties go to the newer fix. It lives in its own file as a plain top-level `getOptimalLocation(locations)` with instrumented tests covering the ordering, staleness, missing-accuracy and threshold-boundary cases. This commit also stops discarding the stored fix: both the batched update callback and the last known location scan now merge it into the selection instead of replacing or skipping it.

2. **Request location from the fused provider alone instead of every provider at once**: previously updates were requested from every provider at once (gps, network, fused, passive), duplicating what fused already combines and forcing the GNSS engine on. The system-app gate from #531 is replaced with `hasProvider(FUSED_PROVIDER)`: availability is the OS's contract, not the caller's privilege. Falls back to gps, network and passive before S, or on the off chance fused is absent on a newer release.

3. **Stop location updates while the activity is paused**: the listener used to stay registered for the process lifetime; now it follows onResume/onPause, and disabling geotagging clears the stored fix.

4. **Stop attaching stale locations to captures**: `getLocation()` returns null for fixes older than 15 minutes, and drops the stored fix so it can't be reused, instead of silently writing an arbitrarily old position into EXIF; the existing location-unavailable message covers that case. `minDistance` goes from 10 m to 0 so stationary users keep receiving fixes (otherwise a still-valid fix ages past any bound).

### CODE REVIEW

File: app/src/main/java/app/grapheneos/camera/App.kt [hunk 1]
```diff
@@ -82,32 +85,25 @@ class App : Application() {
...
-                    // Compare their accuracy instead of time if the difference is below
-                    // threshold
-                    if (location.accuracy > optimalLocation.accuracy) {
-                        optimalLocation = location
-                    }
-                }
-            }
-        }
-        return optimalLocation
+    private fun List<Location?>.getOptimalLocation(): Location? {
```
Comment (RankoR) [inline app/src/main/java/app/grapheneos/camera/App.kt]: This should be properly covered by tests, including edge cases. To do it, would be better to extract it to a top-level function (ideally - in a separate file).

Also, in general we should avoid extension functions unless they are purely private. `getOptimalLocation(locations: List<Location?>): Location?` would be better.

File: app/src/main/java/app/grapheneos/camera/App.kt [hunk 2]
```diff
@@ -126,37 +122,60 @@ class App : Application() {
...
+                0f,
                 locationListener
             )
         }
     }
 
+    // The fused provider combines the underlying providers itself, so use it
+    // alone when it's available (GrapheneOS implements it in the OS; elsewhere
+    // it may be backed by Play services)
+    private val locationProviders: List<String>
```
Comment (RankoR) [inline app/src/main/java/app/grapheneos/camera/App.kt:124]: Pre-S should also use `LocationManager.PASSIVE_PROVIDER`

File: app/src/main/java/app/grapheneos/camera/App.kt [hunk 3]
```diff
@@ -126,37 +122,60 @@ class App : Application() {
...
+        dropLocationUpdates()
+        location = null
+    }
+
+    fun getLocation(): Location? {
+        val location = this.location ?: return null
+        val ageMs = TimeUnit.NANOSECONDS.toMillis(
+            SystemClock.elapsedRealtimeNanos() - location.elapsedRealtimeNanos
+        )
+        return if (ageMs > MAX_LOCATION_AGE) null else location
```
Comment (RankoR) [inline app/src/main/java/app/grapheneos/camera/App.kt]: `this.location = null` is needed here in case if `ageMs > MAX_LOCATION_AGE`

File: app/src/main/java/app/grapheneos/camera/App.kt [hunk 4]
```diff
@@ -82,32 +85,25 @@ class App : Application() {
...
+        // Among the locations that aren't significantly older than the newest
+        // one, pick the most accurate. getAccuracy() is a radius in meters, so
+        // a smaller value is more accurate; a location without accuracy
+        // information is considered least accurate, and ties go to the newer one
+        return locations.filter {
+            val ageDifferenceMs =
+                TimeUnit.NANOSECONDS.toMillis(newest.elapsedRealtimeNanos - it.elapsedRealtimeNanos)
+            ageDifferenceMs <= STALE_LOCATION_THRESHOLD
+        }.minWithOrNull(compareBy(
+            { if (it.hasAccuracy()) it.accuracy else Float.MAX_VALUE },
```
Comment (RankoR) [inline app/src/main/java/app/grapheneos/camera/App.kt]: Hard to read, chaining like `{ -it.elapsedRealtimeNanos }.compareBy<Location> { ... }.thenByDescending { it.elapsedRealtimeNanos }` would be better

**Review by RankoR — CHANGES_REQUESTED** [review]
It will also require a rebase after https://github.com/GrapheneOS/Camera/pull/665 merged

**Review by RankoR — APPROVED** [review]
